<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>MDTS Workflow</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">MDTS Workflow</h1>



<div id="overview" class="section level2">
<h2>Overview</h2>
<p>This package provides the necessary infrastructure to take as raw trio bam files to produce de novo deletion calls.</p>
</div>
<div id="raw-data" class="section level2">
<h2>Raw Data</h2>
<p>The example data used in this package is available as the data package <code>MDTSData</code>. It includes:</p>
<ol style="list-style-type: decimal">
<li>15 simulated bam files organized in 5 trios using the software <code>capsim</code></li>
<li>A pedigree file denoting the trio kinships and the file paths to the raw bam files</li>
<li>A bw file that includes mappability information</li>
</ol>
<p>This process also requires the <code>BSgenome.Hsapiens.UCSC.hg19</code> package. Slighly processed versions of the data is included with <code>MDTS</code>.</p>
</div>
<div id="calculating-mdts-bins" class="section level2">
<h2>Calculating MDTS Bins</h2>
<p>Our method uses empirically derived bins for calculating coverage of targeted seqeucning data. These bins can be significantly different from the probe design file. This portion of the vignette requires using raw data from <code>MDTSData</code>. The primary function responsbile for calculating the MDTS bins is <code>calcBins()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(MDTS); <span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)

<span class="co"># Using the raw data from MDTSData</span>
<span class="kw">setwd</span>(<span class="kw">system.file</span>(<span class="st">&quot;data&quot;</span>, <span class="dt">package=</span><span class="st">&quot;MDTSData&quot;</span>)) 

<span class="co"># Importing the pedigree file that includes information on where to locate the raw bam files</span>
pD =<span class="st"> </span><span class="kw">pData</span>(<span class="st">&quot;pData.ped&quot;</span>)

<span class="co"># Information on the GC content and mappability to estimate GC and mappability for the MDTS bins</span>
genome =<span class="st"> </span>BSgenome.Hsapiens.UCSC.hg19; map_file =<span class="st"> &quot;chr22_100mer_map.bw&quot;</span>

<span class="co"># This command now subsets 5 samples to determine MDTS bins</span>
bins =<span class="st"> </span><span class="kw">calcBins</span>(pD, <span class="dt">n=</span><span class="dv">5</span>, <span class="dt">rl=</span><span class="dv">100</span>, <span class="dt">med=</span><span class="dv">100</span>, <span class="dt">min=</span><span class="dv">5</span>, genome, map_file)</code></pre></div>
</div>
<div id="calculating-coverage-of-mdts-bins" class="section level2">
<h2>Calculating coverage of MDTS bins</h2>
<p>Now that we have a set of MDTS bins, we can proceed to calculate the number of reads that fall into these bins for the entirety of our sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts =<span class="st"> </span><span class="kw">calcCounts</span>(pD, bins, <span class="dt">rl=</span><span class="dv">100</span>)</code></pre></div>
<p>The MDTS bins, raw counts, and pedigree files are included with <code>MDTS</code> and can be loaded as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setwd</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="dt">package=</span><span class="st">&quot;MDTS&quot;</span>))
<span class="kw">load</span>(<span class="st">'bins.RData'</span>)
<span class="kw">load</span>(<span class="st">'counts.RData'</span>)
<span class="kw">load</span>(<span class="st">'pD.RData'</span>)</code></pre></div>
<p>The pedigree file is of form:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(pD)</code></pre></div>
<pre><code>##     subj_id family_id father_id mother_id gender        bam_path
## 3      F1_1        F1      F1_2      F1_3      1 F1_1_sorted.bam
## 6      F1_3        F1         0         0      2 F1_3_sorted.bam
## 9      F1_2        F1         0         0      1 F1_2_sorted.bam
## 12     F2_1        F2      F2_2      F2_3      1 F2_1_sorted.bam
## 15     F2_3        F2         0         0      2 F2_3_sorted.bam
## 6.1    F2_2        F2         0         0      1 F2_2_sorted.bam</code></pre>
<p>The MDTS bins are</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bins</code></pre></div>
<pre><code>## GRanges object with 1654 ranges and 3 metadata columns:
##          seqnames               ranges strand | median_count
##             &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;numeric&gt;
##      [1]    chr22 [50004903, 50006812]      * |       180.37
##      [2]    chr22 [50044645, 50045624]      * |       100.15
##      [3]    chr22 [50045625, 50046049]      * |       100.15
##      [4]    chr22 [50046050, 50046465]      * |       100.04
##      [5]    chr22 [50046466, 50046833]      * |       100.17
##      ...      ...                  ...    ... .          ...
##   [1650]    chr22 [51209086, 51209261]      * |       100.49
##   [1651]    chr22 [51209262, 51209409]      * |       100.56
##   [1652]    chr22 [51210121, 51210304]      * |       100.28
##   [1653]    chr22 [51210707, 51210835]      * |       100.01
##   [1654]    chr22 [51213617, 51213844]      * |        100.3
##                         GC       mappability
##                  &lt;numeric&gt;         &lt;numeric&gt;
##      [1]  0.37434554973822 0.988758726059142
##      [2] 0.461224489795918                 1
##      [3] 0.482352941176471                 1
##      [4] 0.459134615384615                 1
##      [5] 0.494565217391304 0.953351449423834
##      ...               ...               ...
##   [1650] 0.335227272727273 0.769886363636364
##   [1651] 0.439189189189189 0.888513513513513
##   [1652] 0.516304347826087 0.777173913043478
##   [1653] 0.465116279069767 0.856589147286822
##   [1654] 0.609649122807018 0.754385964912281
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>The count matrix where each column is a sample and each row is a bin:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(counts)</code></pre></div>
<pre><code>##        F1_1   F1_3   F1_2   F2_1   F2_3   F2_2   F3_1   F3_3   F3_2   F4_1
## [1,] 190.15 155.14 213.33 227.96 156.58 180.37 212.26 195.62 132.87 166.57
## [2,] 100.15  91.83 109.40 101.43 122.23  90.42  96.66 155.46  94.87 102.48
## [3,] 102.23  89.00  82.16  87.30 100.15  65.89  85.24  82.48 101.68  88.46
## [4,] 100.68  88.84  96.84 110.35  97.74 109.02 111.00 100.04  89.40  82.17
## [5,] 103.85  75.92  80.84  78.85 117.51  98.52  67.77 100.17  98.84  79.99
## [6,]  97.29  94.67  85.68  72.51 104.83  81.41  76.75 109.91 100.14  97.36
##        F4_3   F4_2   F5_1   F5_3   F5_2
## [1,] 173.27 188.87 176.67 183.73 187.07
## [2,]  72.05 107.33 106.82  78.05  98.87
## [3,]  91.23  73.36  92.64 115.60  95.04
## [4,] 105.65  86.45 112.34  84.98  98.75
## [5,]  93.96  79.18 105.77  88.20  92.91
## [6,]  67.55  79.01 113.02 105.91  97.32</code></pre>
</div>
<div id="normalizing-counts-and-calculating-minimum-distance" class="section level2">
<h2>Normalizing counts and calculating Minimum Distance</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mCounts =<span class="st"> </span><span class="kw">normalizeCounts</span>(counts, bins)</code></pre></div>
<pre><code>## [1] &quot;Log Transforming Counts&quot;
## [1] &quot;GC Adjust&quot;
## [1] &quot;Mappability Adjust&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">md =<span class="st"> </span><span class="kw">calcMD</span>(mCounts, bins, pD)</code></pre></div>
</div>
<div id="circular-binary-segmentation-and-deletion-calling" class="section level2">
<h2>Circular Binary Segmentation and Deletion Calling</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cbs =<span class="st"> </span><span class="kw">segmentMD</span>(md, bins)</code></pre></div>
<pre><code>## Analyzing: F1_1 
## Analyzing: F2_1 
## Analyzing: F3_1 
## Analyzing: F4_1 
## Analyzing: F5_1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">denovo =<span class="st"> </span><span class="kw">denovoDeletions</span>(cbs, mCounts, bins)</code></pre></div>
<pre><code>## [1] &quot;Selecting Ccndidate deNnvo deletions&quot;
## [1] &quot;Filtering out families with probable sequencing failure&quot;
## [1] &quot;Calculating problematic bins&quot;
## [1] &quot;Filtering candidates by problematic bins&quot;</code></pre>
<p>The final output is a single detected de novo deletion in family F3:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">denovo</code></pre></div>
<pre><code>## GRanges object with 1 range and 2 metadata columns:
##       seqnames               ranges strand |         m       famid
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;character&gt;
##   [1]    chr22 [50806919, 50810977]      * |   -0.9213        F3_1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
